pkg/chartsync/git.go

```diff
@@ -0,0 +1,407 @@
+package chartsync
+
+import (
+	"context"
+	"path/filepath"
+	"sync"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/weaveworks/flux/git"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/client-go/tools/cache"
+
+	"github.com/fluxcd/helm-operator/pkg/apis/helm.fluxcd.io/v1"
+	clientset "github.com/fluxcd/helm-operator/pkg/client/clientset/versioned"
+	lister "github.com/fluxcd/helm-operator/pkg/client/listers/helm.fluxcd.io/v1"
+	"github.com/fluxcd/helm-operator/pkg/status"
+)
+
+// Condition change reasons
+const (
+	ReasonGitNotReady = "GitNotReady"
+	ReasonGitCloned   = "GitRepoCloned"
+)
+
+// ReleaseQueue is an add-only `workqueue.RateLimitingInterface`
+type ReleaseQueue interface {
+	AddRateLimited(item interface{})
+}
+
+// GitConfig holds the configuration for git operations.
+type GitConfig struct {
+	GitTimeout      time.Duration
+	GitPollInterval time.Duration
+}
+
+// GitChartSource holds the source references to a Helm chart in git,
+// and a minimal working clone.
+type GitChartSource struct {
+	sync.Mutex
```

Comment:
Exposing the lock externally makes it hard to reason about concurrency issues. Also it's not obvious (to me at least) when should one hold the lock. During any action on Export? 

Maybe we can provide the same concurrency guarantees through a more intuitive/encapsulated abstraction. For instance, is it possible to provide the same guarantees through the actions performed on `GitChartSource`?

Reply:
One should hold the lock for the duration one works with the source, as the `Export` has a relationship with all other fields in the structure. Providing multiple methods is thus not an option.

The only way I can think of encapsulating this is to create a callback method, which acquires the lock and defers the unlock. I may however be short sighted in options/possibilities.